packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -219,8 +223,13 @@ class FlutterNativeAssetsBuildRunnerImpl implements FlutterNativeAssetsBuildRunn
   );
 
   @override
-  Future<bool> hasPackageConfig() {
-    return fileSystem.file(packageConfigPath).exists();
+  Future<bool> hasPackageConfig() async {
+    // WARNING: We do not use [fileSystem] here because the code in
```

Comment:
We should consider whether `package:native_assets_builder` should then provide this method instead of here. But that requires releasing a new version of the package and rolling it.

Reply:
Indeed. The code doesn't pass the "symmetry" principle :)

Reply:
Ideally `package:native_assets_builder` would take in a file system so it can be mocked out in tests. Given the comment [here](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/io.dart#L19) I don't think we want to make exceptions to our policy of not allowing `dart:io` imports in the tool and ideally dependencies shouldn't be able to take a code path that results in us accessing the real file system for any reason while a tool test has a memory file system setup.

@andrewkolos and/or @christopherfujino might have their own opinions on this since they have more experience with this.

Reply:
> Ideally `package:native_assets_builder` would take in a file system so it can be mocked out in tests

When discussing with the Flutter team earlier, we decided that passing in a file system and process manager into a dependency and then relying on that doesn't scale. (`package:native_assets_builder` does process calls _and_ uses the file system, just passing in a file system will make all process calls expecting files being there fail.) Any process call needs to be stubbed in a test with the expected output (stdout/stderr/file system), and if the dependency changes what process calls it does, the virtual process manager outputs need to be updated on rolls of the dependency. I believe that for this reason we decided having a `MockNativeAssetsBuilder` that does not delegate to `package:native_assets_builder`.

> I don't think we want to make exceptions to our policy of not allowing `dart:io` imports in the tool

👍

Reply:
> Any process call needs to be stubbed in a test with the expected output (stdout/stderr/file system), and if the dependency changes what process calls it does, the virtual process manager outputs need to be updated on rolls of the dependency. I believe that for this reason we decided having a MockNativeAssetsBuilder that does not delegate to package:native_assets_builder.

:+1:

> I don't think we want to make exceptions to our policy of not allowing dart:io imports in the tool

:+1:

Reply:
Since everyone seems to favor the inconistency in the same class (one method uses `dart:io` while another one uses `FileSystem`), I'm closing this PR.

Reply:
At the risk of restarting a debate. After some more discussion the issue is that 

> I believe that for this reason we decided having a `FakeFlutterNativeAssetsBuildRunner` that does not delegate to `package:native_assets_builder`.

doesn't hold true. All existing hermetic unit tests that are not native assets related, don't use the `FakeFlutterNativeAssetsBuildRunner`. This means they are exercising `FlutterNativeAssetsBuildRunnerImpl` instead. This seems wrong. Hermetic unit tests should be fully hermetic, so they shouldn't by default execute an external dependency that might (or might not) be using dart IO and process calls directly.

The alternative would be to insert `FakeFlutterNativeAssetsBuildRunner` in all hermetic unit tests in PRs which expand the code paths that exercise `package:native_assets_builder` transitively. This seems not scalable. E.g. you'd have to do this for all fakes/mocks in all tests. Maybe that's why there used to be `globals`? But those got banned from using.

If anyone got some insight on what the design philosophy is here, I'd love to learn. (Because now it seems that our hermetic unit tests are not hermetic at all when it comes to dependencies, only when those dependencies are explicitly mocked.)